drm-howto/modeset-atomic.c

```diff
@@ -0,0 +1,1114 @@
+		iter->b = next_color(&iter->b_up, iter->b, 5);
+		buf = &iter->bufs[iter->front_buf ^ 1];
+		for (j = 0; j < buf->height; ++j) {
+			for (k = 0; k < buf->width; ++k) {
+				off = buf->stride * j + k * 4;
+				*(uint32_t*)&buf->map[off] =
+					(iter->r << 16) | (iter->g << 8) | iter->b;
+			}
+		}
+	}
+
+	/* initial modeset on all iterputs */
+	flags = DRM_MODE_ATOMIC_ALLOW_MODESET;
+	ret = drmModeAtomicCommit(fd, req, flags, NULL);
+	if (ret < 0)
+		fprintf(stderr, "modeset atomic commit failed, %d\n", errno);
+
+	return ret;
+}
+
+/*
+ * modeset_draw() changes. If we got here, the modeset already occurred. So in
+ * the loop that calls for modeset_draw_out() we actually perform the atomic
+ * commit that leads to the page-flips.
+ */
+
+static void modeset_draw(int fd)
+{
+	int ret;
+	fd_set fds;
+	time_t start, cur;
+	struct timeval v;
+	drmEventContext ev;
+	struct modeset_out *iter;
+
+	/* init variables */
+	srand(time(&start));
+	FD_ZERO(&fds);
+	memset(&v, 0, sizeof(v));
+	memset(&ev, 0, sizeof(ev));
+	/* Set this to only the latest version you support. Version 2
+	 * introduced the page_flip_handler, so we use that. */
+	ev.version = 2;
+	ev.page_flip_handler = modeset_page_flip_event;
+
+	/* draw on all outputs */
+	for (iter = output_list; iter; iter = iter->next) {
+		iter->r = rand() % 0xff;
+		iter->g = rand() % 0xff;
+		iter->b = rand() % 0xff;
+		iter->r_up = iter->g_up = iter->b_up = true;
+
+		/* here we have the atomic commit that leads to page-flips */
+		modeset_draw_out(fd, iter);
+	}
+
+	/* wait 5s for VBLANK or input events */
+	while (time(&cur) < start + 5) {
+		FD_SET(0, &fds);
+		FD_SET(fd, &fds);
+		v.tv_sec = start + 5 - cur;
+
+		ret = select(fd + 1, &fds, NULL, NULL, &v);
+		if (ret < 0) {
+			fprintf(stderr, "select() failed with %d: %m\n", errno);
+			break;
+		} else if (FD_ISSET(0, &fds)) {
+			fprintf(stderr, "exit due to user-input\n");
+			break;
+		} else if (FD_ISSET(fd, &fds)) {
+			drmHandleEvent(fd, &ev);
+		}
+	}
+}
+
+/*
+ * modeset_cleanup() stays the same.
+ */
+
+static void modeset_cleanup(int fd)
+{
+	struct modeset_out *iter;
+	drmEventContext ev;
+	int ret;
+
+	/* init variables */
+	memset(&ev, 0, sizeof(ev));
+	ev.version = DRM_EVENT_CONTEXT_VERSION;
+	ev.page_flip_handler = modeset_page_flip_event;
+
+	while (output_list) {
+		/* remove from global list */
+		iter = output_list;
+		output_list = iter->next;
+
+		/* if a page-flip is pending, wait for it to complete */
+		iter->cleanup = true;
+		fprintf(stderr, "wait for pending page-flip to complete...\n");
+		while (iter->pflip_pending) {
+			ret = drmHandleEvent(fd, &ev);
+			if (ret)
+				break;
+		}
+
+		/* restore saved CRTC configuration */
+		if (!iter->pflip_pending)
+			drmModeSetCrtc(fd,
+				       iter->saved_crtc->crtc_id,
+				       iter->saved_crtc->buffer_id,
+				       iter->saved_crtc->x,
+				       iter->saved_crtc->y,
+				       &iter->conn_id,
+				       1,
+				       &iter->saved_crtc->mode);
```

Comment:
CRTC state should not be restored, see https://gitlab.freedesktop.org/wayland/weston/-/commit/c1d0f477bd18deb7a605ed88d0ffa46cdd806157

If you have fbcon, that will restore the state. If you have some other KMS programs, it will set its own modes. If you have nothing, your last picture will remain on screen.

Not restoring also allows seamless switching to the next KMS program (if fbcon does not strike in between).

However, if you were using planes other than primary planes, those would be good to disable.

Having a comment here explaining what to restore and what not would be nice.

Reply:
Hmm, but does fbcon restore anything, if you don't manipulate the VT state...

Reply:
Have you tried without this restore code?

Reply:
I've tried without this code now, and it works. The difference is that the text that was being displayed by the VT vanishes (since we are not restoring it). But I suppose this is better than take the risk pointed in https://gitlab.freedesktop.org/wayland/weston/-/commit/c1d0f477bd18deb7a605ed88d0ffa46cdd806157

But to be honest I'm confused about this:
> Having a comment here explaining what to restore and what not would be nice.

What else could be restored?

Reply:
What do you mean the text vanishes? Does it never come back?

What to restore and what not is actually a good question. If you want to make the quitting fast, you can leave all CRTCs running, connectors associated, and FBs (linear modifier) on the primary planes. All other planes should be shut down (but in this example you use none), and other KMS properties (you set none here) should probably be restored or reset, particularly GAMMA, DEGAMMA, and CSC. I've heard that fbcon fails to restore GAMMA.

I think the comment I'm looking for would be just saying that it's ok leave CRTCs running with only the primary plane in use, but everything else should be turned off or reset. Whoever takes over KMS next will program their own stuff.

Reply:
I just asked on #dri-devel and got the reply that, yeah, if you don't set up the VT then just closing the DRM device should be enough for fbcon to take over and restore itself. But it was also mentioned, that some drivers might forget to implement that. Anyway, fbcon should restore itself at least when it prints something, provided that there is no DRM-master on the device and that the VT is not in KD_GRAPHICS mode.

You never touch the VT, so it is not in KD_GRAPHICS.

Reply:
Does the text come back when it vanishes?

Reply:
It does not come back. What was being shown by `stdout` vanishes. Here's the output [with the restore code](https://user-images.githubusercontent.com/59930603/84049988-824b3c00-a983-11ea-91b0-f2d5506fce52.png) and [without the restore code](https://user-images.githubusercontent.com/59930603/84050035-955e0c00-a983-11ea-98e5-0ae4e3aea0a3.png).

Reply:
But the new text appears, so fbcon is fine. No restoring needed.

We could speculate about what happens to the old text that disappears, but it doesn't matter. I think it would be a kernel issue anyway.

Reply:
That makes sense. Done!